crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -676,25 +633,61 @@ where
         // nullifier is being revealed by the tx
         // 4. The transaction must correctly update the note commitment tree
         // in storage with the new output descriptions
+        // Checks on the transparent bundle, if present
         if !(self.valid_spend_descriptions_anchor(&shielded_tx)?
             && self.valid_convert_descriptions_anchor(&shielded_tx)?
             && self.valid_nullifiers_reveal(keys_changed, &shielded_tx)?
-            && self.valid_note_commitment_update(&shielded_tx)?)
+            && self.valid_note_commitment_update(&shielded_tx)?
+            && validate_transparent_bundle(
+                &shielded_tx,
+                &mut changed_balances,
+                &mut transparent_tx_pool,
+                epoch,
+                conversion_state,
+                &mut signers,
+            )?)
         {
             return Ok(false);
         }
 
-        // Checks on the transparent bundle, if present
-        if !self.validate_transparent_bundle(
-            &shielded_tx,
-            changed_balances,
-            &mut transparent_tx_pool,
-            epoch,
-            conversion_state,
-        )? {
-            return Ok(false);
+        // Ensure that every account for which balance has gone down has
+        // authorized this transaction
+        for (addr, pre) in changed_balances.pre {
+            if changed_balances.post[&addr] < pre && addr != masp_address_hash {
+                signers.insert(addr);
+            }
+        }
+
+        // Ensure that this transaction is authorized by all involved parties
+        for signer in signers {
+            let Some(signer) = changed_balances.decoder.get(&signer) else {
+                return Ok(false);
+            };
+            let public_keys_index_map =
+                crate::account::public_keys_index_map(&self.ctx.pre(), signer)?;
+            let threshold = crate::account::threshold(&self.ctx.pre(), signer)?
+                .unwrap_or(1);
+            let max_signatures_per_transaction =
+                crate::parameters::max_signatures_per_transaction(
+                    &self.ctx.pre(),
+                )?;
+            let mut gas_meter = self.ctx.gas_meter.borrow_mut();
+            if tx_data
+                .verify_signatures(
```

Comment:
I'm not completely sure about this. I tend to believe that this kind of validation should be up to the other vps and not the masp. Also this implies that we'd validate signatures more than once which I don't believe is necessary or we might look for signatures which are not there (someone might come up with a vp that does not requires signatures, bad idea but you know...). In general I'd remove this entire `signers` thing

Reply:
You're right, I think this a weakness/limitation of this approach. In the past, we assumed that the inner `Transaction` was the "same" as the outer `Tx`, so if the other VPs accepted the `Tx` then we could assume that the inner `Transaction` was also accepted. But now that the inner `Transaction` can be completely different from the outer `Tx`, I'm not sure how to validate `b` or `a-b` without either directly checking the signatures or somehow calling the other VPs on them...

Reply:
The vps of the accounts involved should be called anyway right? And the accounts that saw a decrease in their balance will check the signature. If we want to be extra safe maybe we can just call `insert_verifier` in the loop, this way we enforce that the appropriate vps will verify the tx and we'll leave the signature check (if any) up to them

Reply:
Ignore the last comment, this is resolved per discussion